encode.go

```diff
@@ -537,6 +551,21 @@ func encPanic(err error) {
 	panic(tomlEncodeError{err})
 }
 
+func marshalDecode(rv reflect.Value) reflect.Value {
```

Comment:
You're already calling `rv.Interface().(Marshaler)` in the caller; pass in a `Marshaler` rather than a `reflect.Value` and avoid the duplicate work.